### PR TITLE
Sidecar template

### PR DESCRIPTION
**Pull request by jnsgruk**
Early work to make the default template a sidecar charm. Starting this early with the hope we might make this a default when Juju 2.9.0 releases into `stable`. 

- Updated `metadata.yaml` to include `bases`, `containers` and `resources`
- Added a default `.gitignore` that excludes the `build/` directory and any `*.charm` objects
- Updated tests for charmcraft init to remove tests for `series`
- Updated the README to change deployment instructions for template charm
- Added example unit tests for Pebble-based charm

Template can be built and deployed with

```bash
$ mkdir test-charm; cd test-charm; python3 -m charmcraft init
$ charmcraft build
$ juju deploy ./test-charm.charm --resource httpbin-image=kennethreitz/httpbin
```

### CODE REVIEW

**Comment by jnsgruk** [comment]
Added a basic unit test for the `pebble_ready` event. In order for it to actually run, you can replace `ops` in the generated `requirements.txt` file with: `https://codeload.github.com/canonical/operator/zip/fd0617ad1473c28f3819ccc74c567f24533bb73d`, which points to a commit in [this PR](https://github.com/canonical/operator/pull/515) that adds basic testing support.

**Comment by jnsgruk** [comment]
Hi all!

This is ready for review now. I've removed the use of `container_pebble_ready` for now (this [PR](https://github.com/canonical/operator/pull/516)), but will make a seperate PR once it's merged into `ops`.

Thanks!

File: charmcraft/commands/init.py [hunk 1]
```diff
@@ -106,17 +104,11 @@ def run(self, args):
         if not re.match(r"[a-z][a-z0-9-]*[a-z0-9]$", args.name):
             raise CommandError("{} is not a valid charm name".format(args.name))
 
-        if args.series is None:
-            series = "[kubernetes]  # TEMPLATE-TODO: change to an Ubuntu series if not using k8s"
-        else:
-            series = yaml.dump(args.series.split(","), default_flow_style=True)
```
Comment (facundobatista) [inline charmcraft/commands/init.py:112]: "series" should also be removed as the arguments of the command, right?

File: charmcraft/templates/init/metadata.yaml.j2 [hunk 2]
```diff
@@ -5,5 +5,18 @@ description: |
   TEMPLATE-TODO: fill out the charm's description
 summary: |
   TEMPLATE-TODO: fill out the charm's summary
-series: {{ series }}
+bases:
+  - name: ubuntu
+    channel: 20.04/stable
```
Comment (facundobatista) [inline charmcraft/templates/init/metadata.yaml.j2:10]: Should we add a "TEMPLATE-TODO" around these values? or as they are for the charm (and not the "payload") it's just fine to leave them as is?

**Review by facundobatista — CHANGES_REQUESTED** [review]
Great work! I annotated some details. Thanks!

File: README.md [hunk 3]
```diff
@@ -88,15 +95,15 @@ Archive:  test-charm.charm
    812617                     84 files
 ```
 
-Now, if you have a Kubernetes cluster with the Juju OLM accessible you can
-directly `juju deploy <test-charm.charm>` to the cluster. You do not need to
-publish your operator on Charmhub, you can pass the charmed operator file around
-directly to users, or for CI/CD purposes.
+Now, if you have a Kubernetes cluster with the Juju OLM accessible you can issue
+`juju deploy <test-charm.charm> --resource httpbin-image=kennethreitz/httpbin`.
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply README.md:99]: This is just required during development (when the charm likely isn't uploaded yet).

Though this has made me realise a typo, should be `./test-charm.charm`.

File: README.md [hunk 4]
```diff
@@ -105,10 +112,9 @@ Use `charmcraft login` and `charmcraft logout` to sign into Charmhub.
 
 ## Charmhub name registration
 
-You can register operator names in Charmhub with `charmcraft register
-<name>`. Many common names have been reserved, you are encouraged to discuss
-your interest in leading or collaborating on a charmed operator in [Charmhub
-Discourse](https://discourse.charmhub.io/).
+You can register operator names in Charmhub with `charmcraft register <name>`.
+Many common names have been reserved, you are encouraged to discuss your interest
+in leading or collaborating on a charmed operator in [Charmhub Discourse](https://discourse.charmhub.io/).
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply README.md:117]: Yep, fixed in next commit :)

File: charmcraft/commands/init.py [hunk 5]
```diff
@@ -22,10 +22,8 @@
 import re
 from datetime import date
 
-import yaml
-
 from charmcraft.cmdbase import BaseCommand, CommandError
-from charmcraft.utils import make_executable, get_templates_environment
+from charmcraft.utils import get_templates_environment, make_executable
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply charmcraft/commands/init.py:26]: Ah, I think this was an autoformat from vscode... I'll revert

File: charmcraft/templates/init/metadata.yaml.j2 [hunk 6]
```diff
@@ -5,5 +5,18 @@ description: |
   TEMPLATE-TODO: fill out the charm's description
 summary: |
   TEMPLATE-TODO: fill out the charm's summary
-series: {{ series }}
+bases:
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply charmcraft/templates/init/metadata.yaml.j2:8]: Will remove - thanks

File: charmcraft/templates/init/src/charm.py.j2 [hunk 7]
```diff
@@ -15,27 +15,55 @@ develop a new k8s charm using the Operator Framework:
...
-from ops.main import main
 from ops.framework import StoredState
+from ops.main import main
 from ops.model import ActiveStatus
 
 logger = logging.getLogger(__name__)
 
 
-class {{ class_name }}(CharmBase):
+class {{class_name}}(CharmBase):
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply charmcraft/templates/init/src/charm.py.j2:25]: Reverted - thanks

File: charmcraft/templates/init/metadata.yaml.j2 [hunk 8]
```diff
@@ -5,5 +5,18 @@ description: |
...
+  - name: ubuntu
+    channel: 20.04/stable
+
+# TEMPLATE-TODO: replace with containers for your workload (delete for non-k8s)
+containers:
+  httpbin:
+    resource: httpbin-image
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply charmcraft/templates/init/metadata.yaml.j2:15]: Not sure I follow? Are you saying we should just remove the `TEMPLATE-TODO` piece?

File: tests/commands/test_init.py [hunk 9]
```diff
@@ -20,12 +20,10 @@
 from argparse import Namespace
 
 import pytest
-import yaml
-
 from charmcraft.cmdbase import CommandError
 from charmcraft.commands.init import InitCommand
 from charmcraft.utils import S_IXALL
-from tests.test_infra import pep8_test, get_python_filepaths, pep257_test
+from tests.test_infra import get_python_filepaths, pep8_test, pep257_test
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply tests/commands/test_init.py:26]: *curses format-on-save*. Fixed :)

File: charmcraft/commands/init.py (hunk 1 above)
Comment (jnsgruk) [reply charmcraft/commands/init.py:112]: Yep - fixed. Thanks.

File: charmcraft/templates/init/metadata.yaml.j2 (hunk 2 above)
Comment (jnsgruk) [reply charmcraft/templates/init/metadata.yaml.j2:10]: I would have said fine to leave as they are sensible defaults for the charm, but I've removed `bases` entirely now :)

**Review by facundobatista — CHANGES_REQUESTED** [review]
All good, almost all done. I just remembered about the `completion.bash` file: you need to remove the `--series` option from the command, there. Thanks!

**Comment by jnsgruk** [comment]
@facundobatista Good catch. Fixed :) Thanks.

**Review by facundobatista — APPROVED** [review]
Great, thanks!

File: charmcraft/templates/init/requirements.txt.j2 [hunk 10]
```diff
@@ -1 +1 @@
-ops
+ops >= 1.2.0
```
Comment (sergiusens) [inline charmcraft/templates/init/requirements.txt.j2:1]:
```suggestion
ops >= 1.2.0

```

**Review by sergiusens — APPROVED** [review]
Looks good, personally I am not a fan of boilerplate I need to remove every time and would much prefer editor autocomplete for yaml through an lsp-server here.

File: charmcraft/templates/init/requirements.txt.j2 (hunk 10 above)
Comment (jnsgruk) [reply charmcraft/templates/init/requirements.txt.j2:1]: Agreed, I think that sounds like a nicer way to go for the future, but this should serve as a nice example in the mean time. Thanks :)

**Pull request merged**